### PR TITLE
Message editor

### PR DESCRIPTION
**Pull request by acrap5**
Adds some GET and POST routes used as a backend for the CWD Character Message Editor. Additionally, adds a function for processing input from the frontend to update the DB.

https://trello.com/c/IHVo5vzm/39-cwd-confirm-create-new-api-routes-for-editing-messages

### CODE REVIEW

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: Are you okay with not abbreviating `probability`?

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: No need to prefix with `parsed`, as `processed` implies it.  I'd name these `processed.message` and `processed.probability`

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: Grammar

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: This is just string interpolation, so let's bring `view.gauges.${req.params.index - 1}.messages` out into its own variable.  It will make it more concise.

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: Any desire to set the message to an object and not have to split fields?

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: Why does this one not set `probability` too?

**Review by Sammidysam — CHANGES_REQUESTED** [review]
Last comment is the worrisome one.  We want probability to be adjustable in a view's messages object too.

**Comment by Sammidysam** [comment]
This is simply a code review and not QA review.  Yesterday during the call, this did not seem to work when run on my computer.  I will test it today to see what's up - await that call as well before moving toward merging.

**Comment by Sammidysam** [comment]
Testing Results
---------------

Gauge messages: wonderful!
View messages: I'm getting errors back when I hit Update.  Nothing seems to be updating here.

Please take a look at the view messages functionality before moving toward merging.

File: src/glyphs.js
Comment (acrap5) [reply src/glyphs.js]: Yup

**Comment by acrap5** [comment]
Comments addressed and functionality breaking bugs have been fixed. Should be ready to ship now.

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: ``` [`${path}.text`] ``` may be more elegant / fun.

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: Intentional to leave in code?

Comment (acrap5) [reply src/glyphs.js]: oopsy

**Comment by Sammidysam** [comment]
@acrap5 Would it be okay to add the password feature before shipping?  I will also make a card for myself to check this over during office hours tomorrow too.

**Comment by acrap5** [comment]
@Sammidysam Yeah of course. I made the comment about it being ready before the meeting and that discussion. I'll take care of that before or at office hours tomorrow.

**Comment by acrap5** [comment]
Password feature: frontend sends the password with the request. Backend then salts the received (plaintext) password with a string that was randomly generated, then hashes the combined string using SHA256 hash. This hash is checked against the hardcoded password, and if it matches then the database will be updated.

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js:8]: Did this getting added to `package.json` too?  We need to make sure that dependencies are all installed when a new user builds this.

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: Can we add an else statement to respond something saying that the password is invalid if it is?

Also, is this the hardcoded password?  This is a public repository, so we will have to change it and load it from a secret file if it is.  Loading from a file will be really easy on a server as opposed to a web client where we are on the other end.

ALSO, we need to use strict equality, I think.

**Comment by Sammidysam** [comment]
I'm going to leave this as changes requested because my guess is that we have the hardcoded password in our code.  We cannot leave passwords exposed in public code.

Anyway, I have not researched secure password retrieval enough to know if this is the standard way of doing it or not, but this sounds familiar enough and I trust your research.

Just let me know if we have exposed the password?

**Comment by Sammidysam** [comment]
I went ahead and added the `js-sha256` module.  In the future, when you need to add a dependency to the project you would just run the following in project root: `npm install js-sha256`

**Comment by Sammidysam** [comment]
A couple issues upon testing:
https://github.com/EnvironmentalDashboard/citywide-dashboard.alpha/blob/master/cwd/app.js#L172-L177
Note that we change our behavior when the provided probability is an array.  Therefore, in order to have it not be attached to data, we need the data to be provided as an integer.  Currently, it seems that we provide an array of size 1.  That will make the views expect data which will cause an error.

Screen Shot 2020-04-06 at 2 18 30 PM
I am not able to edit the view message probability anymore, as it tries to update a gauge's probability rather than the view's message.

Both of these I noticed in code before noticing them here, so if you follow along my code review you should be able to find them without issue.

File: src/glyphs.js
Comment (acrap5) [reply src/glyphs.js:8]: This slipped my mind, thanks for catching and taking care of it.

File: src/glyphs.js
Comment (acrap5) [reply src/glyphs.js]: Now responds 'invalid password' if so, let me know if you want something more than this.

Yes and no, will address this in a separate comment.

You're right, thanks for catching that, fixed.

**Comment by acrap5** [comment]
What's hardcoded is the SHA256 hash of the actual password plus a 'salt' at the end. Realistically, this cannot be decoded (the purpose of hash being one way, vs two-way encryption) and so the actual password cannot (feasibly) be found from what's hardcoded in here. Since this hardcoded hash is checked against "processed.pass", which is salted and hashed in the "processMessageRequest" function before being passed along to the pull request, I don't really see any feasible way that what's hardcoded here could be used to bypass this password measure.

TLDR; I think this system is good enough for now, while obviously the ideal eventual goal is to incorporate the user role (?) system so that only authorized users can make changes (or even access the page).

Let me know if you disagree with any of this or still have concerns.

**Comment by acrap5** [comment]
So you're saying that when updating view messages we should send only an integer, and not an array of length 1?

Also I have no issues with updating the view message probability. I suspect that it's not actually an issue with what you think it is, but rather because of the issue with the "id" vs "ObjectID" in the database that we discussed earlier. I'll prioritize making that change.

File: src/glyphs.js
Comment (Sammidysam) [reply src/glyphs.js]: Could you check some of the other responses present in this API for how they respond errors?  Right now you return only a string when there is an invalid password, but since this is a JSON API we should return some JavaScript object.  I think I have it modeled in other routes, but let me know if you cannot find what I am referencing.

**Comment by Sammidysam** [comment]
First comment: I don't disagree - I just have not read up on security and all in quite some time so I don't really have sufficient knowledge to challenge what you are saying.  I trust you here.  Besides, we can definitely store backups of the database when I get my backup server rolling, and that will help to safeguard if anyone did try super hard to bypass this all.

Second comment: Yes - an array implies that we have data attached to the glyph.  See the `app.js` code that I linked to in my big comment up above.  So therefore, since a view has no data attached, its probability for messages needs to be only an integer.

**Comment by acrap5** [comment]
What's the point of making it a list of strings in this context?

**Comment by Sammidysam** [comment]
Some API responses may have multiple errors and we want to provide a generic format that will allow us to not have to adapt how we handle errors on the frontend to the API path used.

**Comment by acrap5** [comment]
Most recent commit should now have everything working as wanted.

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js:239]: So I am not at all saying you need to do this, but the original idea was that this call would become an external function.  You could provide this function with an ID, a path, and a processed message request, and it would then do the actual database logic.  The reasoning is that then if we added a variable to a message, we would edit it in one place, not two.

So, you are free to implement that idea if you want, but I still feel bad about not being clear so do not want to at all force it.  It was simply a clean code idea.

File: src/glyphs.js
Comment (Sammidysam) [reply src/glyphs.js]: This question is actually still relevant, as the `processMessageRequest` function handles making sure we don't add extra fields from the request, so we can trust the processed object.

**Comment by acrap5** [comment]
This path only gets fired when a gauge doesn't have any messages attached to it yet. Per your own words, a gauge should only have a messages array if it has at least one message attached. So when a adding a first message to a gauge this path gets fired to create the array and then the messages can be added to it and updated as usual.

File: src/glyphs.js
Comment (Sammidysam) [inline src/glyphs.js]: `result => result` is more concise / prettier, if you would like.

Comment (Sammidysam) [reply src/glyphs.js]: And actually, you could just delete the `.then` line in this case.

**Comment by Sammidysam** [comment]
Did you add a `push` message route for gauges?  I don't think I see it, and my Postman was not able to send it.  I think we need this too - should be a quick change.

**Comment by acrap5** [comment]
Sorry, didn't get a notification for this comment so I missed it until now. The current `push` message route is for gauges. Views don't have a `push` message route, I'll add one now.

**Comment by Sammidysam** [comment]
No worries with the notification not firing.  I'll wait until I get an answer on the last comment I just made before working on testing again, as maybe my Postman setup will need to be adjusted depending on how the requests are expected to come in.

**Comment by Sammidysam** [comment]
@acrap5 I just wanted to ping the above question, as the message editor is almost fully done and all.

**Pull request merged**
